packages/sources/canton-functions/src/transport/canton-data.ts

```diff
@@ -56,24 +55,57 @@ export class CantonDataTransport extends SubscriptionTransport<BaseEndpointTypes
     params: RequestParams,
   ): Promise<AdapterResponse<BaseEndpointTypes['Response']>> {
     const providerDataRequestedUnixMs = Date.now()
+    const url = params.url
+    const templateId = params.templateId
+    const choice = params.choice
 
-    const contracts = await this.cantonClient.queryContracts({
-      templateIds: [params.templateId],
-    })
+    let contractId: string
+    let contract: any
+
+    // If contractId is provided, use it directly
+    if (params.contractId) {
+      contractId = params.contractId
+    } else {
+      // Query contracts using contractFilter
+      if (!params.contractFilter) {
+        throw new AdapterInputError({
+          message: 'Either contractId or contractFilter must be provided',
+          statusCode: 400,
+        })
+      }
 
-    if (!contracts || contracts.length === 0) {
-      throw new AdapterInputError({
-        message: `No contracts found for template ID '${params.templateId}'`,
-        statusCode: 404,
+      const contracts = await this.cantonClient.queryContractsByTemplate(url, {
+        templateIds: [templateId],
+        filter: String(params.contractFilter),
       })
+
+      if (!contracts || contracts.length === 0) {
+        throw new AdapterInputError({
+          message: `No contracts found for template ID '${templateId}' with the provided filter`,
+          statusCode: 404,
+        })
+      }
+
+      // Find the latest contract by createdAt
+      contract = this.findLatestContract(contracts)
```

Comment:
I think each eturned contract may have a flag associated with it, which says where it's "ACTIVE" or "ARCHIVED" - it might be useful to check this and log an error